### PR TITLE
Monoids of endofunctions

### PR DESCRIPTION
**Pull request by avekens**
As mentioned in #3782 and proposed by @benjub, the monoid of endofunctions on a set ` A `  can serve as example for a monoid with several left inverses but no right inverse (requested by @savask).

To prove this, I provide a definition `` ( EndoFMnd ` A ) `` for a monoid of endofunctions  on a set ` A `  (see ~ df-efmnd and ~efmnd), analogously to the symmetric groups `` ( SymGrp ` A ) `` (see ~df-symg and ~symgval). I extracted several parts of proofs for ` SymG `, which could be reused in the proofs for ` EndoFMnd `. Furthermore, I added some general theorems about monoids (in main set.mm) which I used in proofs for ` EndoFMnd `.

The main results of this PR are the proofs
*  that `` ( EndoFMnd ` A ) `` are actually monoids (see ~efmndmnd), 
* that there are subsets of the base set of `` ( EndoFMnd ` NN0 ) `` which are monoids (see ~smndex1mnd) with a different identity element (see ~smndex1n0mnd), so that they are not submonoids (see ~nsmndex1). See also the remark in ~submnd0.
* that there is an element in `` ( EndoFMnd ` NN0 ) `` which has no right invere (see ~smndex2dnrinv) but arbitrary many left inverses (see ~smndex2dlinvh). This is the formalization of @benjub's proposal. 
* that `` ( SymGrp ` A ) `` is a submonoid of `` ( EndoFMnd ` A ) `` (see ~symgsubmefmnd)

By the way, I just saw that the last item was on @digama0's TODO list:   

> Make SymGrp a subgroup of EndoMnd (the TopMnd of functions X --> X)

Details:

Main set.mm:
* theorems ~iresn0n0, ~iotan0 added
* theorem ~ 0mp0 added (some proofs are already shortened using this theorem, but there may be additional proofs using ~ mpo0 which can be shortened. Unfortunately, this cannot be done by the minimize script).
* theorems ~mpo0v, ~mgmsscl,  ~sgrpidmnd, ~issubmndb, ~mndissubm, ~resmndismnd, ~0subm added
* ~idresperm moved up
* lemma ~symggrplem added
* proofs of ~grpsubfval , ~symgplusg, ~symggrp  shortened

AV's mathbox:
* Definition of  monoids of endofunctions on set and related basic theorems added
* Example of a monoid contained in another monoid not being a submonoid (because it has a different identity element): ~smndex1mnd and ~nsmndex1
* Example of a monoid having an element with infinitely many left inverses but no right inverse: ~smndex2dlinvh and  ~smndex2dnrinv

### CODE REVIEW

File: set.mm [hunk 1]
```diff
@@ -741438,6 +741624,644 @@ corresponds to the definition of an (infinite) product in [Lang] p. 5,
...
+      @( The monoid of endofunctions on a pair is consisting of the identity,
+         the transposition and two mappings to the same element.  (Contributed
+         by AV, 18-Feb-2024.) @)
+      efmnd2bas @p |- ( ( I e. V /\ J e. W )
+                       -> B = ( { { <. I , I >. , <. J , J >. } ,
+                                { <. I , J >. , <. J , I >. } } u.
+                                { { <. I , I >. , <. J , I >. } ,
+                                { <. I , J >. , <. J , J >. } } ) ) @=
+        ? @.
+$)
```
Comment (tirix) [inline set.mm:741903]: This proof is probably quite laborious. It is a good simple example but I'm not sure it is necessary.

File: set.mm [hunk 2]
```diff
@@ -741438,6 +741624,644 @@ corresponds to the definition of an (infinite) product in [Lang] p. 5,
...
+  ${
+    submefmndbi.g $e |- M = ( EndoFMnd ` A ) $.
+    submefmndbi.b $e |- B = ( Base ` M ) $.
+    submefmndbi.0 $e |- .0. = ( 0g ` M ) $.
+    $( Characterization of submonoids of the monoid of endofunctions on a set
+       ` A ` .  (Contributed by AV, 17-Feb-2024.) $)
+    submefmndbi $p |- ( A e. V -> ( S e. ( SubMnd ` M )
+                       <-> ( S C_ B /\ .0. e. S /\ ( M |`s S ) e. Mnd ) ) ) $=
+      ( wcel cmnd csubmnd cfv wss cress co w3a wb efmndmnd eqid issubm2 syl ) A
+      EJDKJCDLMJCBNFCJDCOPZKJQRADEGSBCUCDFHIUCTUAUB $.
```
Comment (tirix) [inline set.mm:741916]: Ok, but this seems like a direct application of `issubm2`, meaning this is valid for _any_ monoid.

**Review by tirix — APPROVED** [review]
Very nice!
I believe the general part about the monoid of endomorphisms should be moved to main.
The specific part about `NN0` might be interesting in the example sections, too.
Do you also plan to prove that it is a topological monoid?

File: set.mm (hunk 1 above)
Comment (avekens) [reply set.mm:741903]: Yes, we would need a theorem similar to hash2prd:

```
    hash4qdd $p |- ( ( P e. V /\ ( # ` P ) = 4 )
                   -> ( ( ( ( W e. P /\ X e. P ) /\ ( Y e. P /\ Z e. P ) ) 
                            /\ ( ( W =/= X /\ W =/= Y /\ W =/= Z ) 
                              /\ ( X =/= Y /\ X =/= Z ) /\ Y =/= Z ) )
                          -> P = { W ,  X , Y , Z } ) ) $=
```

File: set.mm (hunk 2 above)
Comment (avekens) [reply set.mm:741916]: You are right, this theorem does not give much additional value. I will remove it and replace its (single) usage in ~idresefmnd by ~issubm2.

**Comment by avekens** [comment]
> Do you also plan to prove that it is a topological monoid?

This would be the analog of ~symgtgp, wouldn't it? ~symgtgp has a quite long proof, but maybe I'll have a try...

**Comment by tirix** [comment]
> > Do you also plan to prove that it is a topological monoid?
> 
> This would be the analog of ~symgtgp, wouldn't it? ~symgtgp has a quite long proof, but maybe I'll have a try...

I admit I did not check, but I assume that the exact same proof can be followed, and then `symgtgp` can be obtained by `submtmd` and the fact that the symmetric group is a submonoid of the monoid of the endomorphisms, which you already proved.

**Comment by avekens** [comment]
> > > Do you also plan to prove that it is a topological monoid?
> > 
> > 
> > This would be the analog of ~symgtgp, wouldn't it? ~symgtgp has a quite long proof, but maybe I'll have a try...
> 
> I admit I did not check, but I assume that the exact same proof can be followed, and then `symgtgp` can be obtained by `submtmd` and the fact that the symmetric group is a submonoid of the monoid of the endomorphisms, which you already proved.

It was easy to prove that the monoid of endofunctions on a set ` A ` is a topological monoid (see ~efmndtmd) by transforming a part of the proof for ~symgtgp. But it is not easy (or maybe even not possible) to replace the corresponding part of of the proof for ~symgtgp by using ~efmndtmd, because we do not have 

```
( SymGrp ` A ) = ( ( EndoFMnd ` A ) |`s ( Base ` ( SymGrp ` A )  ) )
```

(by definition, not only the base sets are different, but also the group operations (slot ``( +g ` ndx )``)).

**Comment by avekens** [comment]
Maybe it would be useful to change the definition of SymGrp to
```
SymGrp = ( x e. _V |-> { <. ( Base ` ndx ) ,  { h | h : x -1-1-onto-> x }  >. ,
        <. ( +g ` ndx ) , ( f e. ( x ^m x ) , g e. ( x ^m x ) |-> ( f o. g ) ) >. ,
        <. ( TopSet ` ndx ) , ( Xt_ ` ( x X. { ~P x } ) ) >. } )
```
Then we would have
```
( SymGrp ` A ) = ( ( EndoFMnd ` A ) |`s { h | h : A -1-1-onto-> A } ) 
                       = ( ( EndoFMnd ` A ) |`s ( Base ` ( SymGrp ` A )  ) )
```
and we could reuse a lot of theorems for `EndoFMnd` also for `SymGrp`.

I think I'll open an issue for this.

**Comment by avekens** [comment]
I added the submonoids of injectiv/surjective endofunctions (~sursubmefmnd and ~injsubmefmnd) as proposed by Gérard Lang, allowing for a a shorter proof of ~symgsubmefmnd, also proposed by Gérard Lang (~insubm required).

File: set.mm [hunk 3]
```diff
@@ -741438,6 +741695,716 @@ corresponds to the definition of an (infinite) product in [Lang] p. 5,
...
+       by AV, 17-Feb-2024.) $)
+    idressubmefmnd $p |- ( A e. V -> { ( _I |` A ) } e. ( SubMnd ` G ) ) $=
+      ( wcel cid cres csn c0g cfv csubmnd efmndid sneqd cmnd efmndmnd 0subm syl
+      eqid eqeltrd ) ACEZFAGZHBIJZHZBKJZTUAUBABCDLMTBNEUCUDEABCDOBUBUBRPQS $.
+
+    idresefmnd.e $e |- E = ( G |`s { ( _I |` A ) } ) $.
+    $( The structure with the singleton containing only the identity function
+       restricted to a set ` A ` as base set and the function composition as
+       group operation, constructed by (structure) restricting the monoid of
+       endofunctions on ` A ` to that singleton, is a monoid with a base set as
```
Comment (icecream17) [inline set.mm:742112]:
```suggestion
       endofunctions on ` A ` to that singleton, is a monoid whose base set is a
```

File: set.mm [hunk 4]
```diff
@@ -741438,6 +741695,716 @@ corresponds to the definition of an (infinite) product in [Lang] p. 5,
...
+      ZCKLGZBMGZBNLZCNLZOZTZACDEPVCVFVEVIOZCQLZVEGZCVEUAUBZMGZUCZVKVCCMGVFVQUMA
+      CDEUDVIVEVOCVMVIRZVMRVORUEUFVCVQVKVCVJVQVGVCVHVEVIUGZVIVESGVSVHUHVCVDUIVE
+      VIBSCFVRUJUKVEVIUNULVPVLVGVNVPVGVOBMBVOFUOUPUQURUSUTVAVB $.
+  $}
+
+  ${
+    smndex1ibas.m $e |- M = ( EndoFMnd ` NN0 ) $.
+    smndex1ibas.n $e |- N e. NN $.
+    smndex1ibas.i $e |- I = ( x e. NN0 |-> ( x mod N ) ) $.
+    $( The modulo function ` I ` is an endofunctions on ` NN0 ` .  (Contributed
```
Comment (icecream17) [inline set.mm:742129]:
```suggestion
    $( The modulo function ` I ` is an endofunction on ` NN0 ` .  (Contributed
```

**Review by icecream17 — APPROVED** [review]

**Pull request merged**